src/server.rs

```diff
@@ -136,32 +138,34 @@ fn handle_client(tcp_stream: Arc<Mutex<TcpStream>>, clients_set: &Arc<Mutex<Hash
     }
 
     let binding = client_deps.to_string();
-    let container_dependencies_lock = container_dependencies.lock().unwrap();
-
-    let dependencies: Vec<&str> = if binding.is_empty() {
-        let dependencies_lock = container_dependencies_lock.get(&client_id).unwrap();
-        dependencies_lock.iter().map(|x| x.as_str()).collect()
+    let dependencies: Vec<String> = if binding.is_empty() {
+        container_dependencies
+        .lock().unwrap()
+        .get(&client_id)
+        .cloned()
+        .unwrap_or_default()
     } else {
-        binding.split(':').collect()
+        binding.split(':').map(str::to_string).collect()
```

Comment:
Do we need the extra allocations here to clone the original `&str` values into the `Vec<String>`? I don't see `dependencies` being used mutably anywhere below, it looks like we can use `Vec<&str>`.

Reply:
@snprajwal it seems like the `Vec<&str>` here is pointing into the locked HashMap’s Strings, so it forces the lock to be alive as long as the `&str` borrows. This creates the situation where some threads may have to wait for the thread holding the lock before being able to check for dependencies. Cloning here makes it possible to release the lock immediately. 
Is there an approach to keep using `Vec<&str>` and release the lock for other threads you would suggest?

Reply:
Ah right, in that case let's go ahead with cloning. I don't think there's any trivial workaround for this.